qiskit/transpiler/passes/optimization/template_matching/template_substitution.py

```diff
@@ -492,30 +502,76 @@ def _attempt_bind(self, template_sublist, circuit_sublist):
         from sympy.parsing.sympy_parser import parse_expr
 
         circuit_params, template_params = [], []
+        circ_param_str = ""
+        sub_params = {}
 
         template_dag_dep = copy.deepcopy(self.template_dag_dep)
 
+        # add parameters from circuit to circuit_params, as well as a
+        # string containing parameters to check for duplicates from the template
         for idx, t_idx in enumerate(template_sublist):
             qc_idx = circuit_sublist[idx]
             circuit_params += self.circuit_dag_dep.get_node(qc_idx).op.params
-            template_params += template_dag_dep.get_node(t_idx).op.params
+            for param_exp in self.circuit_dag_dep.get_node(qc_idx).op.params:
+                if isinstance(param_exp, ParameterExpression):
+                    circ_param_str += str(param_exp)
+
+        # add parameters from template to template_params, as well as replace
+        # node parameters that have duplicate names in the circuit params
+        # create a dict of sub_params to substitute remaining nodes not in
+        # the template sublist
+        for idx, t_idx in enumerate(template_sublist):
+            node = template_dag_dep.get_node(t_idx)
+            sub_node_params = []
+            for t_param_exp in node.op.params:
+                if isinstance(t_param_exp, ParameterExpression):
+                    for t_param in t_param_exp.parameters:
+                        if t_param.name in circ_param_str:
+                            new_param_name = "".join(
+                                random.choice(string.ascii_lowercase) for i in range(8)
+                            )
+                            sub_params[t_param] = Parameter(new_param_name)
+                            t_param_exp = t_param_exp.assign(t_param, sub_params[t_param])
```

Comment:
This generates a new parameter every time the clash is encountered, which would likely have caused issues if the same parameter needed to be put into a circuit more than once.  Better to create the new parameters only once, and retrieve them from a dictionary on subsequent accesses.  This can be made quite elegant with `collections.defaultdict` - I've done that in a14dc37441f73e8c9630d43b03ba8a1c9de5c757.